Array API support - continuation

mtsokol opened the pull request:
Hi @hameerabbasi,

This PR add a couple of simple Array API functions:
```
    "finfo",  # ported from numpy
    "iinfo",  # ported from numpy
    "isnan",  # the rest is just element-wise functions
    "isinf",
    "isfinite",
    "equal",
    "not_equal",
    "less",
    "less_equal",
    "greater",
    "greater_equal",
    "square",
    "logaddexp",
    "logical_and",
    "logical_or",
    "logical_xor",
    "trunc",
```
Adds `__bool__`, etc. to `Tensor` class.

It also introduces a few modifications to correctly run `array-api-test` suite.

---

The tricky part is `reshape` which we don't have right now but it's required by most of `array-api-tests` suites to even run them.
Maybe one option could be another env var `TEST_MODE` that when set enables temporary implementation (that moves to numpy, reshapes, and moves back to Tensor). Otherwise `NotImplementedError` is raised. WDYT?

hameerabbasi commented on src/finch/tensor.py:
```diff
@@ -188,18 +191,37 @@ def _elemwise_op(self, op: str, other: Optional["Tensor"] = None) -> "Tensor":
+        return self._to_scalar(complex)
+
+    def _to_scalar(self, builtin):
+        if self.size != 1 or self.shape != ():
+            raise ValueError(f"{builtin} can be computed for one-element tensors only.")
```
```suggestion
        if self.ndim != 0:
            raise ValueError(f"{builtin} can be computed for scalar tensors only.")
```

willow-ahrens replied:
Instead, can we check that all dimensions, if present, are of size 1?

i.e. `if all(dim == 1 for dim in self.shape):`

willow-ahrens replied:
Similarly, could we compute the scalar value by indexing all dimensions with 1?

hameerabbasi replied:
That would be in contrast to the Array API and NumPy.

mtsokol replied:
Right, I think we can stick to Array API standard, and support only 0-D arrays, like the suggested change states:
https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__bool__.html
https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__int__.html

NumPy already warns that >0D is deprecated:
```python
In [43]: np.array([0]).__int__()
<ipython-input-43-6afc8c458142>:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  np.array([0]).__int__()
```

willow-ahrens replied:
sounds good!

mtsokol replied:
Done!

hameerabbasi commented on src/finch/tensor.py:
```diff
@@ -587,25 +609,37 @@ def asarray(obj, /, *, dtype=None, format=None):
 
 
+def reshape(x: Tensor, /, shape: tuple[int, ...], *, copy: bool | None = None) -> Tensor:
+    # Only to run array-api-tests that require it for multiple tests. Maybe raise a warning
+    # that it densifies? Or enable when some env var is present?
```
Let's not densify, `scipy.sparse` made that mistake. It was a source for performance cliffs.

willow-ahrens replied:
I'm on it, perhaps we can leave this for testing for now?

mtsokol replied:
Sure! Thank you!

hameerabbasi commented on src/finch/tensor.py:
```diff
@@ -587,25 +609,37 @@ def asarray(obj, /, *, dtype=None, format=None):
-        return Tensor(np.full(shape, fill_value, dtype=jl_dtypes.jl_to_np_dtype[dtype]))
-    raise ValueError(f"{format} format not supported.")
+    else:
+        return Tensor(np.full(shape, fill_value, dtype=dtype))
```
This returns a fully dense tensor. Can we return an empty tensor with the right fill-value?

willow-ahrens replied:
only if the fill value is "special" right? we wouldn't want to compile a ton of kernels on accident.

mtsokol replied:
There are two formats supported by `full` here: `coo` and `dense`. The `else` branch is for `dense` case so I think we should return an actual dense tensor. Or do we want a COO format?

willow-ahrens replied:
in Julia and Matlab we have `spzeros` to explicitly ask for a sparse array of zeros, as opposed to `zeros` which returns a dense array. I think this would be resolved if we simply require the user to type `dense` or `coo` and the else case is an error. That way, the user can't make a dense tensor on accident.

willow-ahrens replied:
Oh, I see, we check it at the top. In that case, I think this is fine.

hameerabbasi replied:
We could run into a performance cliff if we choose the incorrect default either way. I think the right solution would be the middle solution with a dynamic fill-value, treating it as a runtime constant rather than a compile time constant, except for absorbing or identity elements.

willow-ahrens replied:
Finch doesn't support dynamic sparse fill values yet.

willow-ahrens replied:
I could add it with the ISO level?

willow-ahrens replied:
but it's not a common thing to do

hameerabbasi replied:
Hmm. 🤔 Not an easy decision either way. Let's pick any solution here and create a Todo comment and an issue to fix it up later?

willow-ahrens replied:
I'm okay with the code as written, though I think it could be improved to make it clearer that dense is not the default but an intentional user decision.

hameerabbasi replied:
The issue with that is we're introducing finch-specific extensions to the standard, and they're highly unlikely to get used.

willow-ahrens replied:
which extension here is finch-specific? `full`? or the `format` keyword?

mtsokol replied:
I think nothing is finch-specific. `full` follows Array API with and additional keyword only `format` argument that is also present in `full` in Numba backend.

willow-ahrens replied:
If you could change the `else` to an `elif format == dense`, I think I approve this PR.

hameerabbasi replied:
`format` is sparse-specific.

mtsokol replied:
I updated `if` statement and kept the `format` keyword for this PR but if you think it should be dropped as it's sparse-specific, let me know - I will keep only COO then. (but Numba backend also has `format` keyword for these functions and accepts for example "gcxs")

willow-ahrens replied:
No, I think adding a format keyword is becoming a good, uniform way to incorporate sparsity into these functions without changing the signatures too much, let's keep doing it!

hameerabbasi replied:
I would add `format` to `compute`, to minimise the amount of additional API added. Uniformity is good, less API to do the same thing is even better.

willow-ahrens replied:
I think it's very reasonable for a user to want an empty COO matrix full of ones.

willow-ahrens replied:
so they would type `full(1, format=coo)`. You're proposing that they need to do `compute(lazy(full(1)), format=dense)`?

willow-ahrens replied:
or would it be `compute(full(lazy(1)), format="coo")`? I'm not quite sure where the lazy would go in this invocation. Another issue is that we may want to specify intermediate formats for performance reasons, so it's not just the final format that matters. Finally, it would be nice to be able to change the format without dropping into the lazy interface.

hameerabbasi replied:
I think it would be close to `full(1).compute(format[, optional_schedule])` in a lazy context (i.e. inside a function with `@compiled`.

`compute` would force materialisation in any context.

I guess a middle way would be to add it to creation functions only, but every function would be way too much API surface.

willow-ahrens replied:
I'm not sure we can support a lazy version of this function. I agree that adding `format` to all creation functions is a good plan. As a future API issue, we should consider whether to add format to eager-only operations.

hameerabbasi reviewed (changes requested):
We also need tests for everything here.

hameerabbasi commented on src/finch/__init__.py:
```diff
@@ -100,8 +112,15 @@
+    iinfo,
 )
 
+logical_and = bitwise_and
+logical_or = bitwise_or
+logical_xor = bitwise_xor
+logical_not = bitwise_invert
```
These are distinct -- `logical_*` casts to `bool` before any ops. `bitwise_*` performs bit-wise ops with integers.

willow-ahrens replied:
In Finch, logical and is the `Finch.and` function, while bitwise and is `&`

mtsokol replied:
I found:
```
Finch.and
Finch.or
Finch.xor
```
but I think `invert` or `not` is missing, right? (because it's tricky as it changes the fill value?)

willow-ahrens reviewed (changes requested):
Looks good! Just a few tweaks to bitwise ops, the scalar changes, and the one change in `full`

mtsokol commented:
The PR is ready from my side - I added tests and addressed all comments.
One thing though, I see that Windows is failing, because for numpy arrays after reshaping it says that default value fails on `isbits` check in Finch. This is weird as in `reshape` I pass just a numpy array, it passes in other test cases on Windows.

willow-ahrens commented on src/finch/tensor.py:
```diff
@@ -587,25 +615,38 @@ def asarray(obj, /, *, dtype=None, format=None):
+    # TODO: https://github.com/willow-ahrens/Finch.jl/issues/558
+    #       Only to run array-api-tests that require it for multiple tests.
+    #       Must be reimplemented once `reshape` is available in Finch.jl.
+    arr = x.todense()
```
Could we simply add a PerformanceWarning here that we are densifying?

mtsokol replied:
Done!

willow-ahrens reviewed (changes requested):
Let's just add a warning for the temporary reshape. Otherwise LGTM!

mtsokol commented:
Hi @willow-ahrens @hameerabbasi Can we merge this PR? Do you have any other comments?

willow-ahrens reviewed (approved):
Lgtm!

hameerabbasi commented on src/finch/tensor.py:
```diff
@@ -86,7 +87,7 @@ def __init__(
         fill_value: np.number = 0.0,
     ):
-        if isinstance(obj, (int, float, complex, bool)):
+        if isinstance(obj, (int, float, complex, bool, list)):
```
I'd avoid hard-coding list and check for `Iterable`.

mtsokol replied:
But that would be incorrect. We only support iterables that can be passed to `np.array(...)` to create a valid array.

`dict` and `set` are out of support (only `list` and `tuple`):
```python
In [9]: isinstance(set((1,2,3)), Iterable)
Out[9]: True

In [10]: np.array(set((1,2,3)))
Out[10]: array({1, 2, 3}, dtype=object)

In [12]: isinstance({1:2}, Iterable)
Out[12]: True

In [13]: np.array({1:2})
Out[13]: array({1: 2}, dtype=object)
```

hameerabbasi commented on src/finch/tensor.py:
```diff
@@ -897,8 +949,32 @@ def atanh(x: Tensor, /) -> Tensor:
 
-def atan2(x: Tensor, other: Tensor, /) -> Tensor:
-    return x._elemwise_op("atand", other)
+def atan2(x1: Tensor, x2: Tensor, /) -> Tensor:
+    return x1._elemwise_op("atand", x2)
+
+
+def trunc(x: Tensor, /) -> Tensor:
+    return x._elemwise_op("trunc")
+
+
+def square(x: Tensor, /) -> Tensor:
+    return x ** Tensor(2)
+
+
+def logaddexp(x1: Tensor, x2: Tensor, /) -> Tensor:
+    return log(exp(x1) + exp(x2))
+
+
+def logical_and(x1: Tensor, x2: Tensor, /) -> Tensor:
+    return x1._elemwise_op("Finch.and", x2)
+
+
+def logical_or(x1: Tensor, x2: Tensor, /) -> Tensor:
+    return x1._elemwise_op("Finch.or", x2)
+
+
+def logical_xor(x1: Tensor, x2: Tensor, /) -> Tensor:
+    return x1._elemwise_op("Finch.xor", x2)
```
Can we add tests for these?

mtsokol replied:
They are all tested. Please see `test_elemwise_ops_2_args` and `test_elemwise_ops_1_arg` changes.

hameerabbasi reviewed (changes requested):
A few comments to finish up, seems like the larger discussions are resolved.

mtsokol commented on src/finch/dtypes.py:
```diff
@@ -40,6 +43,17 @@
     complex64: np.complex64,
     complex128: np.complex128,
-    bool: np.bool_,
+    bool: builtins.bool,
```
The same change already happens in `tensor.py::638`:
```
if dtype == np.bool_:  # Fails with: Finch currently only supports isbits defaults
    dtype = bool
```
`np.bool_` isn't accepted by Finch, only builtin `bool` is fine.

hameerabbasi replied:
Is this a `Finch.jl` issue or a Julia issue, @willow-ahrens?

mtsokol replied:
@hameerabbasi Can we inspect it in a follow-up work? We already use `np.bool_` -> `builtins.bool` in a few places in `tensor.py`, I don't think it should be addressed here.

hameerabbasi replied:
Sure, I'll create an issue.
